### PR TITLE
android: support down migrations

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt [hunk 1]
```diff
@@ -39,88 +43,131 @@ fun DatabaseErrorView(
...
-      generalGetString(R.string.wrong_passphrase)
-    else
-      generalGetString(R.string.encrypted_database)
-    is DBMigrationResult.Error -> generalGetString(R.string.database_error)
-    is DBMigrationResult.ErrorKeychain -> generalGetString(R.string.keychain_error)
-    is DBMigrationResult.Unknown -> generalGetString(R.string.database_error)
-    null -> "" // should never be here
+
+  @Composable
+  fun migrationsText(ms: List<String>) {
```
Comment (avently) [inline apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt:77]: Composable functions should be started with Uppercase letter (if they not return anything, e.g. non-unit)

File: apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      ChatConsoleItem { withAuth(showCustomModal { it, close -> TerminalView(it, close) }) }
+      SectionDivider()
+      val devTools = remember { mutableStateOf(developerTools.get()) }
+      SettingsPreferenceItem(Icons.Outlined.Construction, stringResource(R.string.settings_developer_tools), developerTools, devTools)
+      SectionDivider()
+      var confirm = remember { mutableStateOf(confirmDBUpgrades.get()) }
+      SettingsPreferenceItem(Icons.Outlined.DriveFolderUpload, stringResource(R.string.confirm_database_upgrades), confirmDBUpgrades, confirm)
+      SectionDivider()
+      InstallTerminalAppItem(uriHandler)
+      SectionDivider()
```
Comment (spaced4ndy) [inline apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt:43]: wouldn't it be rendering right beween section and gray background?
```suggestion
```

File: apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt [hunk 3]
```diff
@@ -39,88 +43,131 @@ fun DatabaseErrorView(
...
-              if (useKeychain) saveAndRunChatOnClick() else runChat(dbKey.value, chatDbStatus, progressIndicator, appPreferences)
-            }
             if (useKeychain) {
-              SaveAndOpenButton(buttonEnabled, saveAndRunChatOnClick)
+              DatabaseKeyField(dbKey, buttonEnabled, ::saveAndRunChatOnClick)
+              SaveAndOpenButton(buttonEnabled, ::saveAndRunChatOnClick)
             } else {
-              OpenChatButton(buttonEnabled) { runChat(dbKey.value, chatDbStatus, progressIndicator, appPreferences) }
+              DatabaseKeyField(dbKey, buttonEnabled) { callRunChat() }
+              OpenChatButton(buttonEnabled) { callRunChat() }
```
Comment (avently) [inline apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt:107]:
```suggestion
              OpenChatButton(buttonEnabled, ::callRunChat)
```

File: apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt [hunk 4]
```diff
@@ -204,6 +251,14 @@ private fun restoreDb(restoreDbFromBackup: MutableState<Boolean>, prefs: AppPref
   }
 }
 
+private fun mtrErrorDescription(err: MTRError): String =
```
[the remark this answers is not in the document]
Comment (avently) [reply apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt:254]: for example, to make function name longer and descriptive

File: apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+@Composable
+fun DeveloperView(
+  m: ChatModel,
+  showCustomModal: (@Composable (ChatModel, () -> Unit) -> Unit) -> (() -> Unit),
+  withAuth: (block: () -> Unit) -> Unit
+) {
+  Column(
+    Modifier.fillMaxWidth(),
+    horizontalAlignment = Alignment.Start
```
Comment (avently) [inline apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt:27]: it's default alignment, no need to specify it by hand

**Review by spaced4ndy — DISMISSED** [review]

File: apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt [hunk 6]
```diff
@@ -39,88 +43,131 @@ fun DatabaseErrorView(
...
-            DatabaseKeyField(dbKey, buttonEnabled) {
-              if (useKeychain) saveAndRunChatOnClick() else runChat(dbKey.value, chatDbStatus, progressIndicator, appPreferences)
-            }
             if (useKeychain) {
-              SaveAndOpenButton(buttonEnabled, saveAndRunChatOnClick)
+              DatabaseKeyField(dbKey, buttonEnabled, ::saveAndRunChatOnClick)
+              SaveAndOpenButton(buttonEnabled, ::saveAndRunChatOnClick)
             } else {
-              OpenChatButton(buttonEnabled) { runChat(dbKey.value, chatDbStatus, progressIndicator, appPreferences) }
+              DatabaseKeyField(dbKey, buttonEnabled) { callRunChat() }
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply apps/android/app/src/main/java/chat/simplex/app/views/database/DatabaseErrorView.kt:106]: don't think you can do it - there is an optional parameter

File: apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt (hunk 5 above)
Comment (epoberezkin) [reply apps/android/app/src/main/java/chat/simplex/app/views/usersettings/DeveloperView.kt:27]: was there before, but can remove

**Pull request merged**